File: src/pages/htmx/interval.ts
import type { APIRoute } from "astro";

const element = (interval: number) => `
    ${interval}
    <input 
        type="hidden"
        name="interval"
        value="${interval}"
    />
`;

function parseNumber(value: string | undefined): number {
    const i = Number(value);
    return isNaN(i) ? 0 : i;
}

export const get: APIRoute = async () => {
    return new Response(element(0), {
        headers: {
            "content-type": "text/html",
        },
    });
};

export const post: APIRoute = async ({ request }) => {
    const formData = await request.formData();
    const interval = parseNumber(formData.get("interval")?.toString());

    return new Response(element(interval + 1), {
        headers: {
            "content-type": "text/html",
        },
    });
};

File: htmx/src/interval.rs
use actix_web::{get, post, web, HttpRequest, HttpResponse};
use leptos::*;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct IntervalForm {
    interval_rust: String,
}

fn element(index: i32) -> String {
    leptos::ssr::render_to_string(move |cx| {
        view! { cx,
            {index}
            <input
                type="hidden"
                name="interval_rust"
                value="{index}"
            />
        }
    })
}

#[get("/interval")]
pub async fn get(_: HttpRequest) -> HttpResponse {
    return HttpResponse::Ok()
        .content_type("text/html; charset=utf-8")
        .body(element(0));
}

#[post("/interval")]
pub async fn post(_req: HttpRequest, info: web::Form<IntervalForm>) -> HttpResponse {
    let counter = info.interval_rust.parse::<i32>().unwrap();

    return HttpResponse::Ok()
        .content_type("text/html; charset=utf-8")
        .body(element(counter + 1));
}
